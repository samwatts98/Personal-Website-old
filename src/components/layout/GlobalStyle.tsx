import { createGlobalStyle } from 'styled-components';

export const THEME_TRANSITION = 'background-color 0.5s, color 0.5s';
export const DESKTOP_MEDIA_QUERY = '@media screen and (min-width: 500px)';

export const GlobalStyle = createGlobalStyle`
  body {
    background-color: ${(props) => props.theme.primaryColour};
    color: ${(props) => props.theme.textColour};
    transition: ${THEME_TRANSITION};
    font-family: 'Poppins', sans-serif;
  }
  nav {
    background-color: ${(props) => props.theme.secondaryColour};
    transition: ${THEME_TRANSITION};
  }
  button, a {
    
    border: none;
    outline: none;
    background: transparent;
    transition: ${THEME_TRANSITION};
    color: ${(props) => props.theme.textColour};
    &:active {
      background-color: ${(props) => props.theme.ternaryColour};
    }
  }
`;
