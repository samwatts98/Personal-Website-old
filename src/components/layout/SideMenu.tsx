import styled, { css } from 'styled-components';
import React, { useState } from 'react';
import { pageLinks, PageLinkProps } from 'components/util/PageRouter';
import { THEME_TRANSITION, DESKTOP_MEDIA_QUERY } from './GlobalStyle';
import { Link } from 'react-router-dom';
import { HasChildren } from 'types/common';
import { FiMenu } from 'react-icons/fi';

type ToggleProps = {
  expanded: boolean;
};

const SideMenuComponent = styled.nav<ToggleProps>`
  display: flex;
  position: fixed;
  padding: 1rem;

  top: 0;
  flex-direction: column;
  width: 100%;
  height: 3rem;

  ${({ expanded }) =>
    expanded
      ? css`
          height: fit-content;
        `
      : css`
          height: 3rem;
          > a {
            display: none;
          }
        `};

  ${DESKTOP_MEDIA_QUERY} {
    width: 3rem;
    height: 100vh;
    transition: width 0.2s ease-in, ${THEME_TRANSITION};

    > a {
      display: flex;
    }

    &:hover {
      width: 15rem;
      .menuItemName {
        opacity: 1;
        transition: opacity 0.2s ease-in 0.2s;
      }
    }
    .menuToggle {
      display: none;
    }
  }
`;

const SideMenuItemComponent = styled(Link)`
  display: flex;
  align-items: baseline;
  background-color: Transparent;
  border: none;
  outline: none;
  border-radius: 0.5rem;
  width: min-content;
  padding: 0 0.5rem;
  transition: transform 0.2s, ${THEME_TRANSITION};
  text-decoration: none;

  .menuItemName {
    padding-left: 1rem;
  }

  ${DESKTOP_MEDIA_QUERY} {
    &:hover {
      transform: translateX(1rem);
    }
    .menuItemName {
      opacity: 0;
      transition: opacity 0.2s;
    }
  }

  svg {
    vertical-align: text-bottom;
  }
`;

const ToggleMenu = styled.button`
  /* margin: auto 0 auto auto; */
  background-color: Transparent;
  border: none;
  outline: none;
  width: min-content;
  height: min-content;
  justify-self: center;
  padding-left: 0.5rem;
  p {
    margin: 0.5rem 0;
  }
`;

function SideMenuItem({ MenuIcon, text, to }: PageLinkProps): React.ReactElement {
  return (
    <SideMenuItemComponent to={to} aria-label={text}>
      <p className="menuIcon">
        <MenuIcon size={'2rem'} />
      </p>
      <h2 className="menuItemName">{text}</h2>
    </SideMenuItemComponent>
  );
}

export function SideMenu({ children }: HasChildren): React.ReactElement {
  const [expanded, setExpanded] = useState(false);
  const handleClick = (): void => {
    setExpanded((e) => !e);
  };
  return (
    <SideMenuComponent expanded={expanded}>
      <ToggleMenu onClick={handleClick} className="menuToggle">
        <p>
          <FiMenu size={'2rem'} />
        </p>
      </ToggleMenu>
      {children}
      {pageLinks.map((link) => (
        <SideMenuItem key={link.text} {...link} />
      ))}
    </SideMenuComponent>
  );
}
