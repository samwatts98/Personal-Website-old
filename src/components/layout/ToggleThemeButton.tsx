import styled from 'styled-components';
import React from 'react';
import { IconType } from 'react-icons/lib';

import { DESKTOP_MEDIA_QUERY, THEME_TRANSITION } from './GlobalStyle';

const ToggleThemeButtonComponent = styled.button`
  position: absolute;
  right: 3rem;
  transition: transform 0.1s ease-in, ${THEME_TRANSITION};

  margin: 0.5rem;
  ${DESKTOP_MEDIA_QUERY} {
    padding: 0.5rem 0;
    width: min-content;
    height: min-content;
    position: static;
  }

  &:active {
    background-color: Transparent;
  }
  &:hover {
    transform: scale(1.4);
  }
`;

type ToggleThemeButtonProps = {
  handleClick: () => void;
  Icon: IconType;
};

export function ToggleThemeButton({ handleClick, Icon }: ToggleThemeButtonProps): React.ReactElement {
  return (
    <ToggleThemeButtonComponent onClick={handleClick}>
      <Icon />
    </ToggleThemeButtonComponent>
  );
}
