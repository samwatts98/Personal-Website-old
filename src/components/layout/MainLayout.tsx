import React from 'react';
import styled from 'styled-components';
import { Light, Dark } from 'components/common/Themes';
import { SideMenu } from 'components/layout/SideMenu';
import { ThemeProvider } from 'styled-components';
import { GlobalStyle, DESKTOP_MEDIA_QUERY } from 'components/layout/GlobalStyle';
import { IconContext } from 'react-icons';
import { ToggleThemeButton } from 'components/layout/ToggleThemeButton';

type MainLayoutProps = {
  children?: React.ReactNode;
};

const MainSection = styled.main`
  display: block;
  margin: 6rem 1rem 2rem 1rem;
  text-align: justify;

  ${DESKTOP_MEDIA_QUERY} {
    margin: 2rem 1rem 2rem 6rem;
  }
`;

const MainLayout = ({ children }: MainLayoutProps): React.ReactElement => {
  const [darkMode, setDarkMode] = React.useState(false);

  const handleThemeChange = () => {
    setDarkMode((dm) => !dm);
  };

  return (
    <ThemeProvider theme={darkMode ? Dark : Light}>
      <IconContext.Provider value={{ size: '2rem' }}>
        <GlobalStyle />
        <SideMenu>
          <ToggleThemeButton handleClick={handleThemeChange} darkMode={darkMode} />
        </SideMenu>
        <MainSection>{children}</MainSection>
      </IconContext.Provider>
    </ThemeProvider>
  );
};

export default MainLayout;
