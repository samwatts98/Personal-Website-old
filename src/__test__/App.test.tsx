import React from 'react';
import renderer from 'react-test-renderer';
import { ToggleThemeButton } from 'components/layout/ToggleThemeButton';
import { FiHome, FiCode } from 'react-icons/fi';

it('renders', () => {
  let icon = FiHome;
  const getIcon = () => {
    icon = icon === FiHome ? FiCode : FiHome;
  };

  const tree = renderer.create(<ToggleThemeButton handleClick={getIcon} Icon={icon} />);
  expect(tree).toMatchSnapshot();
});
'';
