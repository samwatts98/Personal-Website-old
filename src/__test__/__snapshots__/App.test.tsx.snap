// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<button
  className="ToggleThemeButton__ToggleThemeButtonComponent-sc-1p7nc9j-0 dBqCnn"
  onClick={[Function]}
>
  <svg
    fill="none"
    height="1em"
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth="2"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3 9l9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
    />
    <polyline
      points="9 22 9 12 15 12 15 22"
    />
  </svg>
</button>
`;
